source/command-monitoring/tests/unified/redacted-commands.yml

```diff
@@ -177,76 +210,131 @@ tests:
               command: { copydb: { $$exists: false } }
 
   - description: "hello with speculative authenticate"
+    runOnRequirements:
+      - minServerVersion: 4.9
     operations:
       - name: runCommand
         object: *database
         arguments:
           commandName: hello
           command:
-            hello: "private"
-            speculativeAuthenticate: "foo"
-        expectError:
-          isError: true
+            hello: 1
+            speculativeAuthenticate:
+              saslStart: 1
+    expectEvents:
+      - client: *client
+        events:
+          - commandStartedEvent:
+              commandName: hello
+              command:
+                hello: { $$exists: false }
+                speculativeAuthenticate: { $$exists: false }
+          - commandSucceededEvent:
+              commandName: hello
+              reply:
+                # We expect the reply not to be redacted, as the server can't
+                # successfully authenticate us, which means the reply does not
+                # need to be redacted
+                isWritablePrimary: { $$exists: true }
```

Comment:
Hmm confused what's going on here. The Go driver will redact the `reply` field here because this is a `hello` with `speculativeAuthenticate`, `isWritablePrimary` will not exist, and this test will not pass. Why is the redaction conditional on the success of the command?

Reply:
seconding this comment, as the same is true for the node driver

Reply:
`hello` (and legacy hello for the rest of this comment) need to be handled slightly differently. While for all other commands we make the decision to redact based on the command name alone, `hello` only needs to be redacted when it contains sensitive information, i.e. the `speculativeAuthenticate` field. However, the reply to a sensitive command may not be necessarily sensitive (as in this case where we fail speculative authentication and the `hello` reply is as if we hadn't attempted speculative authentication in the first place).

Due to the architecture of the C driver (where I built these tests), we can no longer inspect the original command document when evaluating whether a reply needs to be redacted. Thus, the C driver only redacts replies to `hello` if the reply itself contains a `speculativeAuthenticate` field. While I believe this behaviour is fine and still fulfils the intention of the spec (redacting sensitive information), this behaviour doesn't necessarily apply to other drivers.

I've updated the tests to use `$$unsetOrMatches`: this allows drivers to always redact replies to `hello` when the command document used speculative authentication even when the reply does not contain sensitive information. It also accepts the behaviour of drivers that only redact the reply if it contains sensitive information.

Note that this only applies to `hello` - other replies are always redacted in the C driver. I'll make a second attempt at creating tests that allow successful authentication (e.g. to be able to test the response to `saslStart` and `hello` with speculative authentication), but I believe this to be outside of the scope of this ticket as it will most likely require more changes to the unified test format, which I'd like to avoid at this time.

Reply:
Yep, in the Go driver, the `redacted` flag in a `commandSucceededEvent` is directly inherited from the corresponding `commandStartedEvent`, so the logic isn't granular enough to recognize when a command is security-sensitive but its reply is not. Using `$$unsetOrMatches` makes sense to me.

Reply:
I'm not sure this will work actually, see my [most recent comment](https://github.com/mongodb/specifications/pull/1026#discussion_r658799927).

Reply:
Resolving this comment thread as this issue is addressed in a separate comment.